htsget/htsget.go

```diff
@@ -0,0 +1,224 @@
+				AcceptEncoding string `json:"accept-encoding"`
+				Authorization  string `json:"authorization"`
+			} `json:"headers,omitempty"`
+		} `json:"urls"`
+	} `json:"htsget"`
+}
+
+// Htsget function downloads the files included in the urls_list.txt file.
+// The argument can be a local file or a url to an S3 folder
+func Htsget(args []string) error {
+
+	// Call ParseArgs to take care of all the flag parsing
+	err := helpers.ParseArgs(args, Args)
+	if err != nil {
+		return fmt.Errorf("failed parsing arguments, reason: %v", err)
+	}
+
+	var currentPath string
+	currentPath, err = os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current path, reason: %v", err)
+	}
+
+	if *datasetID == "" || *fileName == "" || *htsgetHost == "" || *publicKeyFile == "" {
+		return fmt.Errorf("missing required arguments, datasetID, fileName, htsgetHost and key are required")
+	}
+
+	config, err := helpers.GetAuth(*configPath)
+	if err != nil {
+		return err
+	}
+
+	//read public key from file
+	publickey, err := os.ReadFile(currentPath + "/" + *publicKeyFile)
+	if err != nil {
+		return fmt.Errorf("failed to read public key, reason: %v", err)
+	}
+	base64publickey := base64.StdEncoding.EncodeToString(publickey)
+
+	// TODO: Add cases for different type of files
+	// i.e. bam files require the /reads/, replace for vcf
+	url := *htsgetHost + "/reads/" + *datasetID + "/" + *fileName
+	if *referenceName != "" {
+		url = url + "?referenceName=" + *referenceName
+	}
+
+	method := "GET"
+
+	client := &http.Client{}
+	req, err := http.NewRequest(method, url, nil)
+	if err != nil {
+		return fmt.Errorf("failed to make request, reason: %v", err)
+	}
+
+	req.Header.Add("Authorization", "Bearer "+config.AccessToken)
+	req.Header.Add("client-public-key", base64publickey)
+
+	res, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to do the request, reason: %v", err)
+	}
+	defer res.Body.Close()
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response, reason: %v", err)
+	}
+
+	var htsgetURLs htsgetResponse
+	err = json.Unmarshal(body, &htsgetURLs)
+	if err != nil {
+		return fmt.Errorf("error unmarshaling response, reason: %v", err)
+	}
+
+	err = downloadFiles(htsgetURLs, config)
+	if err != nil {
+		return fmt.Errorf("error downloading the files, reason: %v", err)
+	}
+	return nil
+
+}
+
+func downloadFiles(htsgeURLs htsgetResponse, config *helpers.Config) (err error) {
+
+	// Create the file in the outdir or current location
+	var currentPath string
+	currentPath, err = os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current path, reason: %v", err)
+	}
+	if outDir != nil && *outDir != "" {
+		currentPath = *outDir
+	}
+	out, err := os.OpenFile(currentPath+"/"+*fileName+".c4gh", os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	//read public key from file
+	publickey, err := os.ReadFile(currentPath + "/" + *publicKeyFile)
```

Comment:
if `outDir` is used, `currentPath` is changed (on line 157) and the path to the key will be assumed to be from that path:
```sh
 ./sda-cli htsget -config [ ... ] -key ../path/folder/keys/c4gh.pub.pem  -outdir pang
Error: error downloading the files, reason: failed to read public key, reason: open pang/../path/folder/keys/c4gh.pub.pem: no such file or directory
```